src/java.base/share/classes/java/util/stream/Gatherers.java

```diff
@@ -0,0 +1,686 @@
+     *
+     * @see java.util.stream.Stream#reduce(Object, BinaryOperator)
+     *
+     * @param initial the identity value for the fold operation
+     * @param folder the folding function
+     * @param <T> the type of elements the returned gatherer consumes
+     * @param <R> the type of elements the returned gatherer produces
+     * @return a new Gatherer
+     * @throws NullPointerException if any of the parameters are null
+     */
+    public static <T, R> Gatherer<T, ?, R> fold(
+            Supplier<R> initial,
+            BiFunction<? super R, ? super T, ? extends R> folder) {
+        Objects.requireNonNull(initial, "'initial' must not be null");
+        Objects.requireNonNull(folder, "'folder' must not be null");
+
+        class State {
+            R value = initial.get();
+            State() {}
+        }
+
+        return Gatherer.ofSequential(
+                State::new,
+                Integrator.ofGreedy((state, element, downstream) -> {
+                    state.value = folder.apply(state.value, element);
+                    return true;
+                }),
+                (state, downstream) -> downstream.push(state.value)
+        );
+    }
+
+    /**
+     * Runs an effect for each element which passes through this gatherer,
+     * in the order in which they appear in the stream.
+     *
+     * <p>This operation produces the same elements it consumes,
+     * and preserves ordering.
+     *
+     * @see #peek(Consumer)
+     *
+     * @param effect the effect to execute with the current element
+     * @param <TR> the type of elements the returned gatherer consumes and produces
+     * @return a new gatherer which executes an effect, in order,
+     *         for each element which passes through it
+     * @throws NullPointerException if the provided effect is null
+     */
+    public static <TR> Gatherer<TR, ?, TR> peekOrdered(
+            final Consumer<? super TR> effect) {
+        Objects.requireNonNull(effect, "'effect' must not be null");
+
+        class PeekOrdered implements Gatherer<TR, Void, TR>,
+                                     Integrator.Greedy<Void, TR, TR> {
+            // Integrator
+            @Override
+            public Integrator<Void, TR, TR> integrator() {
+                return this;
+            }
+
+            // Integrator implementation
+            @Override
+            public boolean integrate(Void state,
+                                     TR element,
+                                     Downstream<? super TR> downstream) {
+                effect.accept(element);
+                return downstream.push(element);
+            }
+        }
+
+        return new PeekOrdered();
+    }
+
+    /**
+     * Runs an effect for each element which passes through this gatherer,
+     * in the order in which they are processed -- which in the case of parallel
+     * evaluation can be out-of-sequence compared to the sequential encounter
+     * order of the stream.
+     *
```

Comment:
How does this differ from Stream::peek?

Reply:
@briangoetz Basically the same as `Stream::parallel().forEach` vs `Stream.parallel().forEachOrdered`.

Reply:
peek and peekOrdered have been removed